### ISSUE TITLE
Add Hebrew support

### ISSUE THREAD

**Issue by AshiVered** [sspanak/tt9#180]
I'm working about it.
(Please note, my fork of Traditional T9, called [Qinboard T9](https://github.com/AshiVered/QinBoard-T9) supports Hebrew. But it is less sophisticated than this great fork)
Also, I'm working on contributing more features  from my fork to this fork.

**Comment by sspanak** [comment]
I will definitely appreciate some help with adding one more language.

So, if you haven't already, please check out the [Adding a New Language guide](https://github.com/sspanak/tt9/blob/master/CONTRIBUTING.md#adding-a-new-language). Since, you have technical knowledge, it should be very easy for you to follow it.

> Also, I'm working on contributing more features from my fork to this fork.

Please, do so in separate pull requests.

**Comment by AshiVered** [comment]
I am trying to compile the code, after adding Hebrew, but I have some problems
![image](https://user-images.githubusercontent.com/76535872/219889278-fe9891ed-16de-4d99-bb50-5b794fbb6583.png)
I am not a very experienced programmer, and I did not find (let's start with the first error) the class "Locale" where the variable "hebrew" is not present.
Is it related to
java.util.Locale
?

**Comment by AF1512** [comment]
I'm not sure but I think in the Language.java file you should write `locale = new Locale("he","HE");` (based on the other Languages.java).

**Comment by AshiVered** [comment]
where? it's all code from Language.java
```
package io.github.sspanak.tt9.languages;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Locale;


public class Language {
	protected int id;
	protected String name;
	protected Locale locale;
	protected int icon;
	protected String dictionaryFile;
	protected int abcLowerCaseIcon;
	protected int abcUpperCaseIcon;
	protected ArrayList<ArrayList<String>> characterMap = new ArrayList<>();
	private final HashMap<Character, String> reverseCharacterMap = new HashMap<>();

	// settings
	protected boolean isPunctuationPartOfWords; // see the getter for more info

	final public int getId() {
		return id;
	}

	final public Locale getLocale() {
		return locale;
	}

	final public String getName() {
		return name;
	}

	final public int getIcon() {
		return icon;
	}

	final public String getDictionaryFile() {
		return dictionaryFile;
	}

	final public int getAbcIcon(boolean lowerCase) {
		return lowerCase ? abcLowerCaseIcon : abcUpperCaseIcon;
	}


	/**
	 * isPunctuationPartOfWords
	 * This plays a role in Predictive mode only.
	 *
	 * Return "true", if you need to use the 1-key for typing words, such as:
	 * "it's" (English), "a'tje" (Dutch), "п'ят" (Ukrainian).
	 *
	 * Return "false" also:
	 * 		- hide words like the above from the suggestions.
	 *		- 1-key would commit the current word, then display the punctuation list.
	 * 			For example, pressing 1-key after "it" would accept "it" as a separate word,
	 * 			then display only: | , | . | ! | ? | ...
	 *
	 * "false" is recommended when apostrophes or other punctuation are not part of the words,
	 * because it would allow faster typing.
	 */
	final public boolean isPunctuationPartOfWords() { return isPunctuationPartOfWords; }


	/************* utility *************/

	private void generateReverseCharacterMap() {
		reverseCharacterMap.clear();
		for (int digit = 0; digit <= 9; digit++) {
			for (String keyChar : getKeyCharacters(digit)) {
				reverseCharacterMap.put(keyChar.charAt(0), String.valueOf(digit));
			}
		}
	}


	public String capitalize(String word) {
		return word != null ? word.substring(0, 1).toUpperCase(locale) + word.substring(1).toLowerCase(locale) : null;
	}

	public boolean isMixedCaseWord(String word) {
		return word != null
			&& (
				(word.length() == 1 && word.toUpperCase(locale).equals(word))
				|| (!word.toLowerCase(locale).equals(word) && !word.toUpperCase(locale).equals(word))
			);
	}

	public ArrayList<String> getKeyCharacters(int key, boolean includeDigit) {
		if (key < 0 || key >= characterMap.size()) {
			return new ArrayList<>();
		}

		ArrayList<String> chars = new ArrayList<>(characterMap.get(key));
		if (includeDigit && chars.size() > 0) {
			chars.add(String.valueOf(key));
		}

		return chars;
	}

	public ArrayList<String> getKeyCharacters(int key) {
		return getKeyCharacters(key, true);
	}

	public String getDigitSequenceForWord(String word) throws InvalidLanguageCharactersException {
		StringBuilder sequence = new StringBuilder();
		String lowerCaseWord = word.toLowerCase(locale);

		if (reverseCharacterMap.isEmpty()) {
			generateReverseCharacterMap();
		}

		for (int i = 0; i < lowerCaseWord.length(); i++) {
			char letter = lowerCaseWord.charAt(i);
			if (!reverseCharacterMap.containsKey(letter)) {
				throw new InvalidLanguageCharactersException(this, "Failed generating digit sequence for word: '" + word);
			}

			sequence.append(reverseCharacterMap.get(letter));
		}

		return sequence.toString();
	}

	@NonNull
	@Override
	public String toString() {
		return name != null ? name : "";
	}
}
```

**Comment by AshiVered** [comment]
not help.
this is my code.
https://github.com/AshiVered/tt9

**Comment by AshiVered** [comment]
@sspanak have you any ideas?

**Comment by sspanak** [comment]
[According to the Oracle](https://www.oracle.com/java/technologies/javase/jdk8-jre8-suported-locales.html), the correct locale is iw-IL. So, I would recommend this:
```java
locale = new Locale("iw","IL");
```

By the way, I just realized Hebrew has no capital letters. This means adding it, will require some more work. AB (capitals) mode needs to be disabled for such languages. But don't worry about this, I'll take care of it. I am making a note for myself, not to forget it.

**Comment by AshiVered** [comment]
>[According to the Oracle](https://www.oracle.com/java/technologies/javase/jdk8-jre8-suported-locales.html), the correct locale is iw-IL. So, I would recommend this:

>locale = new Locale("iw","IL");

@sspanak yes. I do it, but it's not compiled....

**Comment by sspanak** [comment]
@AshiVered, if [this](https://github.com/sspanak/tt9/commit/07f73eb78cc93af61431d2ae5cd32c218f40765a) is the commit in question, then the problem is you have repeated `locale = ...`. Assign it only once and it will work. :wink: 

Also, make sure not to commit the Gradle downgrade. It is OK for testing, but we don't want to use it for building the final APK.

**Comment by AshiVered** [comment]
OK, but now i have this problem
error: cannot find symbol
import io.github.sspanak.tt9.languages.Punctuation;
                                      ^
  symbol:   class Punctuation
  location: package io.github.sspanak.tt9.languages

**Comment by AshiVered** [comment]
finish (with some issues with hebrew lang icon)!
[see this pull request
](https://github.com/sspanak/tt9/pull/190)download built apk (debug)
https://github.com/AshiVered/tt9/releases/tag/0.1

**Comment by sspanak** [comment]
All issues have been resolved and Hebrew seems to be ready to release.

@AshiVered, @AF1512 and anyone else following the issue, you can try out [beta-18.5](https://github.com/sspanak/tt9/releases/tag/hebrew) and let me know if something is wrong. Otherwise, I'll merge it in the following days and it will become part of the upcoming version 19.

**Comment by AF1512** [comment]
Seems to work great! Not sure if I should open another issue but maybe there should be a toast notification when you change languages (maybe input modes also).

**Comment by DarthFlip** [comment]
Seems to work great!

**Comment by AshiVered** [comment]
Hello! It works great, but there is one problem - because of me.
The database I uploaded is causing problems...
A proper database is attached.
Sorry!
P.S. I need a little help with changing UI colors in [my fork of clam's tt9](https://github.com/AshiVered/QinBoard-T9). can you help me? Thank you!
[he-utf8.txt](https://github.com/sspanak/tt9/files/10876745/he-utf8.txt)
